src/net/server/middleware/tsig.rs

```diff
@@ -0,0 +1,460 @@
+//! TSIG message authentication middleware.
+
+use core::convert::Infallible;
+use core::future::{ready, Ready};
+use core::marker::PhantomData;
+use core::ops::{ControlFlow, DerefMut};
+
+use std::sync::Arc;
+use std::vec::Vec;
+
+use futures::stream::{once, Once, Stream};
+use octseq::{Octets, OctetsFrom};
+use tracing::{error, trace, warn};
+
+use crate::base::iana::{Opcode, OptRcode, Rcode, TsigRcode};
+use crate::base::message_builder::AdditionalBuilder;
+use crate::base::wire::Composer;
+use crate::base::{Message, ParsedName, Question, Rtype, StreamTarget};
+use crate::net::server::message::Request;
+use crate::net::server::service::{
+    CallResult, Service, ServiceFeedback, ServiceResult,
+};
+use crate::net::server::util::{mk_builder_for_target, mk_error_response};
+use crate::rdata::tsig::Time48;
+use crate::tsig::{
+    self, KeyName, KeyStore, ServerSequence, ServerTransaction,
+};
+
+use super::stream::{MiddlewareStream, PostprocessingStream};
+
+//------------ TsigMiddlewareSvc ----------------------------------------------
+
+/// TSIG message authentication middlware.
+///
+/// This middleware service validates TSIG signatures on incoming requests, if
+/// any, and adds TSIG signatures to responses to signed requests.
+///
+/// | RFC    | Status  |
+/// |--------|---------|
+/// | [8945] | TBD     |
```

Comment:
Why does it say TBD here?

Reply:
Maybe we just comment it out? It doesn't seem to add anything to the documentation at the moment. Or just change TBD to something like Implemented. What else do we want to do with TSIG?

Reply:
If a quick review of RFC 8945 shows that we've implemented it then indeed we can say it is implemented. This table was really just a reminder to self to check what is implemented and note any omissions. It doesn't have to be in this format. If we implement all of 8945 from a server perspective then we can just say in the descriptive text that it is RFC 8945 compliant and get rid of the table.

Reply:
I think it would be good to do that before merging. Then we don't have to come back to this. I don't think there is much to TSIG processing.

Reply:
Agreed.